Interface for collecting function pointer targets

xbauch opened the pull request:
Motivation: we want `remove_returns` to work even if not preceded by `remove_function_pointers`. For example:

```
int (* const fptbl2[])(void) = {g1, g2};

int func(int id1)
{
  t = fptbl2[id1];  // for this instruction the set is {"g1", "g2"}
  return t;
}
```

This PR exposes the function-pointer-target-collecting part of `remove_function_pointers` to be used in other analyses and the does so in `remove_returns`.

In `remove_returns` use the `possible_fp_targets_map` to be able to remove returns for calls to function pointers. `CALL x=*fp()` should become `CALL *fp(); ASSIGN x = (fp == &f00) ? return_value_f00 : (fp == &f01) ?` ....

The first 3 commits are refactoring/preparation. The fourth introduces the interface. The fifth implements the new `do_function_call` for function pointers.

Then we refactor the `remove_function_pointers` to extract the collect-candidate-functions part and separate the modify-function-call part.

Finally -- now that the file is open -- we also refactor the internal `remove_function_pointers` without any functional change.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Extract do_function_call loop into separate method

No functional change.

Commit:
Refactor do_function_calls

Move the checks for is_function_call and does-return-anything outside; and
pre-compute the is_stub. do_function_call now always does something.

Commit:
Split do_function_call into two for stub/non-stub

Also extract the common behaviour to do_function_calls.

Commit:
Set up the interface for fp_target map

The type and the function to return the map.

petr-bauch committed:
Add do_function_call for function pointers

In case the function id is dereference we grab the set of possible targets (via
the possible_fp_targets_map) and pass it to the new do_function_call method.

There we build a cond_expr by iterating over the possible targets and adding a
case: (f==target ? target_return_value) to it. This cond_expr then becomes the
right-hand-side of the new assignment. We also have to dead all the return
values that were created, see https://github.com/martin-cs/ASVAT/issues/46.

Also building remove_function_pointerst requires messaget

And now so does remove_returnst, hence extending the call sites as well.

petr-bauch committed:
Extract the collection of candidate functions

From remove_function_pointers to a separate function split into three:
1. get_funtion_pointer_targets .. the wrapper (for model/program)
2. refine_call_type .. for incomplete function types
3. try_remove_const_fp .. call remove_const_function_pointer

Then implement a wrapper that returns a map: function-name -> list-of-candidates

petr-bauch committed:
Re-implement remove_function_pointers

Using the new candidate-collecting function.

petr-bauch committed:
Refactor remove_function_pointers

Introducing two new member function:
1. build_new_code .. wraps the collection of function calls and gotos
2. remove_function_pointer_log .. wraps the logging and statistics

Setting the location now takes place inside the loop over candidate functions:
there is no second loop over newly introduced gotos.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 101bf069d5a5f805c3175b413994c67514a637ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108667731
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e0f11726f97a92d9edf17c7069db8a6f8c99c2dd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108669241
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

NlightNFotis committed:
Add tests for listing potential targets and remove/restore returns

And update the unit makefile.

petr-bauch committed:
Assume and assert existence of the right target

And correctly report the case when there are no targets.

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -632,7 +632,7 @@ bool janalyzer_parse_optionst::process_goto_program(const optionst &options)
 
     // remove returns, gcc vectors, complex
-    remove_returns(goto_model);
+    remove_returns(log.get_message_handler(), goto_model);
```
I'm not sure we're entirely consistent with this, but I think we mostly pass a message handler as one of the later parameters. So here I'd swap the two.

xbauch replied:
Well at least for `remove_function_pointers` the `message_handler` is the first parameter. But ok, swapped: https://github.com/diffblue/cbmc/pull/4536/commits/e6cc081a0c172fca1456ddb09b3ad947019a5877.

tautschnig commented on src/goto-programs/remove_returns.cpp:
```diff
@@ -13,17 +13,24 @@ Date:   September 2009
 public:
-  explicit remove_returnst(symbol_table_baset &_symbol_table):
-    symbol_table(_symbol_table)
+  explicit remove_returnst(
```
Are those changes co-ordinated with @kroening? If it's not the case, please do get together to make sure the two of you don't end up working in opposite directions.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -756,7 +756,7 @@ void jbmc_parse_optionst::process_goto_function(
   };
 
-  remove_returns(function, function_is_stub);
+  remove_returns(log.get_message_handler(), function, function_is_stub);
```
Move the message handler to the end of the parameter list.

tautschnig reviewed (commented):
My comments are barely technical, but mostly about process: can I please be reassured that this effort is co-ordinated with the work that @kroening is doing?

martin-cs commented on src/goto-programs/remove_returns.cpp:
```diff
@@ -163,22 +188,46 @@ bool remove_returnst::do_function_calls(
-      function_call.function().id() == ID_symbol,
+      function_call.function().id() == ID_symbol ||
+        function_call.function().id() == ID_dereference,
       "indirect function calls should have been removed prior to running "
```
The reason for the invariant doesn't seem to be correct any more.

xbauch replied:
Modified the comment: https://github.com/diffblue/cbmc/pull/4536/commits/1de8bc629d1e541670db79e13cad8892c94bf02b.

martin-cs commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -420,95 +420,26 @@ void remove_function_pointerst::remove_function_pointer(
-  code_typet call_type=to_code_type(function.type());
-
-  // refine the type in case the forward declaration was incomplete
-  if(call_type.has_ellipsis() &&
```
I'm not quite sure that this behaviour has been carried over with the changes.  Is this handled by one of the new branches?

xbauch replied:
Yes, it was moved to a dedicated method `refine_call_type` which is now called during init-state of `get_function_pointer_targets`.

martin-cs commented on src/goto-programs/remove_function_pointers.cpp:
```diff
@@ -144,6 +144,33 @@ class remove_function_pointerst
+  /// loc1: f1(); goto N+1;
+  /// loc2: f2(); goto N+1;
+  /// ..
+  ///
```
Can we have a `locN+1` at the end for clarity?

xbauch replied:
Done.

martin-cs commented on unit/analyses/remove_function_pointers.cpp:
```diff
@@ -0,0 +1,129 @@
+  // void f(){};
+  symbolt f;
+  f.name = "f";
+  f.mode = ID_C;
```
Would it be an unreasonable ask to have one function which requires the type refinement rules so that `void f(int nargs, ...)` can match `void (*fp) (int, float)` or whatever the rule for ellipsis is?

xbauch replied:
Added: https://github.com/diffblue/cbmc/pull/4536/commits/be87ae5d8b389bedf9f6de0913aacbbf4a729c1d.

martin-cs replied:
Appreciated.  Thank you.

martin-cs commented on src/goto-programs/remove_function_pointers.h:
```diff
@@ -16,12 +16,23 @@ Date: June 2003
+
 // remove indirect function calls
 // and replace by case-split
 void remove_function_pointers(
```
One more thing; can there be a version of `remove_function_pointers` that takes in a `possible_fp_targets_mapt` as that way we have a way of producing more compact instantiations of function pointers.

xbauch replied:
Well, that was fun: https://github.com/diffblue/cbmc/pull/4536/commits/abedd92d69561afdd1909f61b240aaaf15e711f9.

martin-cs replied:
Sorry; that was rather more work than I had thought it would be.  Thanks for doing.

martin-cs reviewed (approved):
Approve subject to changes.

petr-bauch committed:
Swap argument order for remove_returns

so that message handler is not the first.

petr-bauch committed:
Improve comments and fix invariant message

that was not up-to-date.

petr-bauch committed:
Add a unit test for remove function pointers

matching variant function pointer to a complete function.

tautschnig commented:
@kroening @xbauch Is there an ongoing conversation to align the changes the two of you are working on?

xbauch commented:
@tautschnig Well, there is a conversation. @kroening raised the issue that with this PR's changes `remove_return` will effectively subsume `remove_function_pointers`.

petr-bauch committed:
Helper for extracting function identifier

from a function call expression.

petr-bauch committed:
Track the state of function-pointer-removal

on a per-callee basis. To prevent the state for one callee to be used when
removing pointers for another callee.

petr-bauch committed:
Export additional interface for remove-function-pointers

that takes the target map as argument.

To avoid duplication we also refactor the existing functions a bit.

petr-bauch committed:
Fix doxygen

will be squashed.

xbauch commented:
@martin-cs It was a good thing that you suggested a new interface taking the `possible_function_targetst` map because it turned out I wasn't collecting the function ids to index that map with correctly. I'm extracting them here: https://github.com/diffblue/cbmc/pull/4536/commits/c048b5df600c28db2b32d2b6ebe57aa0bc6f5e9d. I also had to start keeping track of the state of the `rfp` on per-callee basis: https://github.com/diffblue/cbmc/pull/4536/commits/759b7f3629df7f7cb5869c0c276b65291a9d86e5.

petr-bauch committed:
Fix broken rebase

will be squashed.

xbauch commented:
@martin-cs and @kroening I've put together an alternative implementation in #4650. It does the same thing but the collection of function pointers is completely separate from remove-function-pointers. I also describe there how remove-returns will differ from remove-function-pointers there. It share some commits to this PR so if you want to read it, I recommend starting here: https://github.com/diffblue/cbmc/pull/4650/commits/5a395352436899ef4ce022b32ec854af15f2be69.

martin-cs commented:
Perhaps it would be useful for me to say why this functionality was desirable.  I would like to be able to run `ait` before function pointer removal.  This seems doable and may even help us get down to just the one abstract interpretation framework.  It will also allow more stateful analysis to produce more compact sets of possible function pointer candidates.  However it would be good to keep function removal as a precondition for `ait` as otherwise we have to mimic the same effect in `ait` and that would be kind of daft.

TGWDB commented:
Closing due to age. If you believe this is erroneous please reopen and update the PR.
